src/SecureStore.CyberArkCCP/Services/HttpClientCache.cs

```diff
@@ -0,0 +1,57 @@
+﻿using System;
+using System.Net.Http;
+using System.Security.Cryptography.X509Certificates;
+using Microsoft.Extensions.Caching.Memory;
+
+namespace UiPath.Orchestrator.Extensions.SecureStores.CyberArkCCP.Services
+{
+    internal class HttpClientCache
+    {
+        private static readonly MemoryCacheOptions _cacheOptions = new MemoryCacheOptions
+        {
+            SizeLimit = 500,
+        };
+
+        private readonly TimeSpan _httpClientExpiration = TimeSpan.FromHours(6);
+
+        private readonly MemoryCache _clients = new MemoryCache(_cacheOptions);
+        private readonly IHttpClientFactory _httpClientFactory;
+        
+        public HttpClientCache(IHttpClientFactory httpClientFactory)
+        {
+            _httpClientFactory = httpClientFactory;
+        }
+
+        public HttpClient Get(string key) =>
+            _clients.Get<HttpClient>(key);
+
+        public HttpClient AddWithCertificate(string key, string certificateAuthorityThumbprint, X509Certificate2 certificate) =>
+            _clients.GetOrCreate(key, e => 
+            {
+                e.Size = 1;
+                e.SlidingExpiration = _httpClientExpiration;
+                return _httpClientFactory.CreateWithCertificate(certificate, certificateAuthorityThumbprint);
+            });
+
+        public void TryRemove(string key)
+        {
+            var removedClient = _clients.Get<HttpClient>(key);
+            if (removedClient == null)
+            {
+                return;
+            }
+            _clients.Remove(key);
```

Comment:
No need to get and then remove. `_clients.Remove` returns the item it removed. You should dispose that.
We need to be careful when working with a class in a thread safe way. Usually consecutive calls to it should not be done for an atomic operation. For example, if between `Get` and `Remove` someone else manages to remove the client and add another one, we would dispose an already disposed item and remove an item that is not meant to be removed.

Reply:
I understand, but `Remove` method does not return the item paired with the key.